### DISCUSSION TITLE
cms sign with MIME information (openssl3.07 version)

### DISCUSSION THREAD

**Discussion by pengzhiyong2023** [openssl/openssl#25960] [Q&A] [answered]
use cms sign cmd, use three-level certificate(ca->intermediate->device),  cmd as blow:
openssl cms -sign -in origianl_data.txt -signer device.pem -inkey devicekey.pem -certfile Intermediate.pem -outform pem -out signed_data.cms

use openssl cms -verify -in signed_data.cms -content origianl_data.txt -CAfile ca.pem -purpose any
It will report the error as blow:
Error reading SMIME Content Info
8012BB86087F0000:error:068000D1:asn1 encoding routines:SMIME_read_ASN1_ex:no content type:crypto/asn1/asn_mime.c:420:

check generated cms file, it missed SMIME content. 

-----BEGIN CMS-----
MIIFsgYJKoZIhvcNAQcCoIIFozCCBZ8CAQExDTALBglghkgBZQMEAgEwCwYJKoZI
hvcNAQcBoIIDFjCCAxIwggH6oAMCAQICCGazNF2HTsMNMA0GCSqGSIb3DQEBCwUA
MDoxCzAJBgNVBAYTAmNuMQswCQYDVQQLEwJjYTELMAkGA1UEAxMCY2ExETAPBgkq
hkiG9w0BCQEWAmNhMB4XDTI0MTExNDA2NDQwMFoXDTI1MTExNDA2NDIwMFowRjEL
MAkGA1UEBhMCY24xDzANBgNVBAsTBmlzc3VlcjEPMA0GA1UEAxMGaXNzdWVyMRUw
EwYJKoZIhvcNAQkBFgZpc3N1ZXIwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK
----
FUyJmMMs+2rx4JHZ3f9SC54KlHkEVfxeP+93v/dySICYWA9De2ZITcO5wbU9WCgq
LoFlFVBmGKt71Mg93vhZMV8Phh8iV89Eeg8y9pS3CNFIj9SYAX0/eYZLOr5Mz3XU
P09Uyyo2aybAlksmzbe9gL+VKtWdVqOyAR2nQc/jEE78JEG4sdsTv6MHddkpCjKX
2GoVWesynwFaQAVWit3xiuWh6ZB73ecxIwxWh9UhBIZ5mSQBKRNO2fqtSlGp1IA1
g0XgxBFulXZ6UIV/A7pnYIhwhNI+EA==
-----END CMS-----

But if use two-level certificate (root->device), generated cms include SMIME 

MIME-Version: 1.0
Content-Disposition: attachment; filename="smime.p7m"
Content-Type: application/pkcs7-mime; smime-type=signed-data; name="smime.p7m"
Content-Transfer-Encoding: base64

MIIFyAYJKoZIhvcNAQcCoIIFuTCCBbUCAQExDTALBglghkgBZQMEAgEwCwYJKoZI
hvcNAQcBoIIDLDCCAygwggIQoAMCAQICCCD/+SsIipPlMA0GCSqGSIb3DQEBCwUA
MEYxCzAJBgNVBAYTAmNuMQ8wDQYDVQQLEwZpc3N1ZXIxDzANBgNVBAMTBmlzc3Vl
cjEVMBMGCSqGSIb3DQEJARYGaXNzdWVyMB4XDTI0MTExNDA2NDUwMFoXDTI1MTEx
--
NhPU+ByboH7wOW4nffM4MBfP+RRePN8ZiWd235DhRFS7AgMBAAGjGjAYMAkGA1Ud
EwQCMAAwCwYDVR0PBAQDAgKEMA0GCSqGSIb3DQEBCwUAA4IBAQB4kqOLvR9jN6uz
wI/TcHhB5O52bxOXPSdME63GpnGZqXDlxCnbj0Gqt68abeIv4E39YUGxNbCAxDP5
nWHZ8MPAmXLGjz1RNLyoK6AhrG+OoP3/NbKMg726V02mriRkbOTdEKzbMAeZK3OD
70udVNZ/GMOZ5ksQcvXM6rc6ctm8f+ksr1934BRokAfS8n3f5ZuNYGS916HoBw+O

**Comment by t8m** [comment, marked as answer]
The output does not have the mime headers in the first case because you are using the `-outform PEM` option.

**Reply by pengzhiyong2023** [reply]
yes, but I use "openssl cms -verify -in signed_data.cms -content origianl_data.txt -CAfile ca.pem -purpose any"   to verify, it will report the error “Error reading SMIME Content Info”.  I missed some parameters?  In addition, in the c++ code, I used PEM_write_bio_CMS_stream to write cms file,  which interface should I  use to parse the cms, I used PEM_read_bio_CMS,  but seems it is deprecated!

**Reply by t8m** [reply]
Yes, please try to add an `-inform PEM` option.
